core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/TransportListTasksAction.java

```diff
@@ -110,4 +150,215 @@ protected void processTasks(ListTasksRequest request, Consumer<Task> operation)
     }
+
+    private class FailOverToTasksIndexListener implements ActionListener<ListTasksResponse> {
+        private final Task task;
+        private final ListTasksRequest request;
+        private final ActionListener<ListTasksResponse> delegate;
+
+        public FailOverToTasksIndexListener(Task task, ListTasksRequest request, ActionListener<ListTasksResponse> delegate) {
+            this.task = task;
+            this.request = request;
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void onResponse(ListTasksResponse listTasksResponse) {
+            if (listTasksResponse.getTasks().isEmpty()) {
+                startGet();
+            } else {
+                delegate.onResponse(listTasksResponse);
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            if (e instanceof NoSuchNodeException) {
+                startGet();
+            }
+            delegate.onFailure(e);
+        }
+
+        private void startGet() {
+            GetRequest get = new GetRequest(TaskResultsService.TASK_RESULT_INDEX, TaskResultsService.TASK_RESULT_TYPE,
+                    request.getTaskId().toString());
+            get.setParentTask(clusterService.localNode().getId(), task.getId());
+            client.get(get, new ActionListener<GetResponse>() {
+                @Override
+                public void onResponse(GetResponse getResponse) {
+                    try {
+                        if (getResponse.isExists()) {
+                            onNonEmptyGetResponse(getResponse);
+                        } else {
+                            // Didn't find the task either way so just return the empty response
+                            delegate.onResponse(new ListTasksResponse(emptyList(), emptyList(), emptyList()));
+                        }
+                    } catch (Exception e) {
+                        delegate.onFailure(e);
+                    }
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    delegate.onFailure(e);
+                }
+            });
+        }
+
+        private void onNonEmptyGetResponse(GetResponse response) throws IOException {
+            if (response.isSourceEmpty()) {
+                throw new IllegalArgumentException("Stored task status for [" + response.getId() + "] didn't contain any source!");
+            }
+            TaskInfo taskInfo;
+            try (XContentParser parser = XContentHelper.createParser(response.getSourceAsBytesRef())) {
+                taskInfo = TaskInfoBuilder.PARSER.parse(parser, () -> ParseFieldMatcher.STRICT).build();
+                delegate.onResponse(new ListTasksResponse(singletonList(taskInfo), emptyList(), emptyList()));
+            }
+        }
+    }
+
+    /**
+     * Builds TaskInfo from persisted copy in the task result index.
+     */
+    public static class TaskInfoBuilder {
```

Comment:
That should probably go to TaskInfo, especially parser that should be definitely somewhere close to the corresponding toXContent method that generates this json.